DRAFT: proposed change to the ast implementation

jessekelly881 opened the pull request:
Fixed type casting. Fixes issue #48 

The _Parser.parse_ function returns a `ParseResult<A>` but it is being cast to `ParseResult<unknown>` in the `encoder` function. The relevant code is:

```js
export interface Parser<A> extends Schema<A> {
  readonly parse: (input: unknown, options?: ParseOptions) => ParseResult<A>
}

export const encode = <A>(
  schema: Schema<A>
): (a: A, options?: ParseOptions) => ParseResult<unknown /* should be A */> => parserFor(schema, "encoder").parse
```

gcanti commented:
> ParseResult<unknown /* should be A */>

In general it's not true

```ts
import { pipe } from "@fp-ts/core/Function"
import * as S from "@fp-ts/schema"
import { parseNumber } from "@fp-ts/schema/data/parser"

expect(S.encodeOrThrow(pipe(S.string, parseNumber))(1)).toEqual("1")
```

jessekelly881 commented:
Ah. I see. The issue is a bit deeper than just a casting issue. The parserFor function isn't actually doing the type transformation, it's just returning Parcer<A> for everything. Have you considered expanding the Schema type so that it can better keep track of the underlying AST? I've played with this for adding proper type checking to the rest function.  

```js
export interface Schema<A, B extends AST.AST> {
  readonly A: (_: A) => A
  readonly ast: B
}

const number: Schema<number, AST.NumberKeyword> = I.number
// etc. 
```

gcanti commented:
> Have you considered expanding the Schema type so that it can better keep track of the underlying AST?

Only one type parameter is needed since a `Schema<A>` is the runtime representation of the TypeScript type `A`

jessekelly881 commented:
@gcanti You could also derive the _A from the AST. I added a test implementation to this pr. There are actually a lot of benefits to doing it this way. I've implemented full function schemas, intersection types, maps between types(preserving type). You could also implement optics between schemas using this implementation. It should also preserve types better between function calls. The only downside that I can see vs the current implementation is with typeLiterals but I think that they are possible with this implementation.

gcanti commented:
I'm not sure I'm following, `encode` cannot have the following signature

```ts
export const encode = <A>(
  schema: Schema<A>
) => (a: A, options?: ParseOptions): ParseResult<A> => ...
```

in any implementation that takes into account a parsing process and not just validation.

In addition, if the implementation only performs validation, then an `encode` function is not even necessary.

By parsing, I mean a transformation like the one I showed above, from string to number, for example.
By validation, I mean just the assertion of a type, for example that an input is a string, and nothing else.
